python/grass/app/data.py

```diff
@@ -162,3 +166,67 @@ def ensure_default_data_hierarchy():
     mapset_path = os.path.join(gisdbase, location, mapset)
 
     return gisdbase, location, mapset, mapset_path
+
+
+class MapsetLockingException(Exception):
+    pass
+
+
+def lock_mapset(mapset_path, force_lock_removal, message_callback):
+    """Acquire a lock for a mapset and return name of new lock file
+
+    Raises MapsetLockingException when it is not possible to acquire a lock for the
+    given mapset either becuase of existing lock or due to insufficient permissions.
+    A corresponding localized message is given in the exception.
+
+    A *message_callback* is a function which will be called to report messages about
+    certain states. Specifically, the funtion is called when forcibly unlocking the
+    mapset.
+
+    Assumes that the runtime is setup (GISBASE).
+    """
+    if not os.path.exists(mapset_path):
+        raise MapsetLockingException(_("Path '{}' doesn't exist").format(mapset_path))
+    if not os.access(mapset_path, os.W_OK):
+        error = _("Path '{}' not accessible.").format(mapset_path)
+        stat_info = os.stat(mapset_path)
+        mapset_uid = stat_info.st_uid
+        if mapset_uid != os.getuid():
+            error = "{error}\n{detail}".format(
+                error=error,
+                detail=_("You are not the owner of '{}'.").format(mapset_path),
+            )
+        raise MapsetLockingException(error)
+    # Check for concurrent use
+    lockfile = os.path.join(mapset_path, ".gislock")
+    locker_path = os.path.join(os.environ["GISBASE"], "etc", "lock")
+    ret = subprocess.run(
+        [locker_path, lockfile, "%d" % os.getpid()], check=False
+    ).returncode
+    msg = None
+    if ret == 2:
+        if not force_lock_removal:
+            msg = _(
+                "{user} is currently running GRASS in selected mapset"
+                " (file {file} found). Concurrent use of one mapset not allowed.\n"
+                "You can force launching GRASS using -f flag"
+                " (assuming your have sufficient access permissions)."
+                " Confirm in a process manager "
+                "that there is no other process using the mapset."
+            ).format(user=Path(lockfile).owner(), file=lockfile)
+        else:
+            message_callback(
+                _(
+                    "{user} is currently running GRASS in selected mapset"
+                    " (file {file} found), but forcing to launch GRASS anyway..."
+                ).format(user=Path(lockfile).owner(), file=lockfile)
+            )
```

Comment:
In https://github.com/OSGeo/grass/pull/4052, I was implicitly asked to not have translated strings changed, because of some amount of caution. So I worked to keep the scope of that PR strictly limited to having the strings collected. The translation files aren't kept updated in the repo.

So, if we are to accept to change the source string (including the placeholder types) here, then there shouldn't be any problems at all that would be worth holding up #4052 any longer, as it will suffer the same consequences.

Reply:
I think the idea is to have not have the strings changed unnecessarily. If the previous old string is not good or if the new code requires adjustment, then the string should be changed. I think here, it will go to another file, so that also makes the string different even if it would be the same text.

Let's not discuss #4052 here, but as my comment there suggests, I don't see any problem with that PR at all.

Reply:
Ok, I'm fine with not extending here, I just wanted us to be consequential with our reviews in between PRs